text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1827 @@
+
+## Why migrate away from `?Sized`?
+[why-migrate-away-from-sized]: #why-migrate-away-from-sized
+
+`?Sized` is frequently regarded as confusing for new users and came up in the
+[prior art][prior-art] as a reason why new `?Trait` bounds were not seen as desirable
+(see [rfcs#2255][issue_more_implicit_bounds]). Furthermore, it isn't clear that `?Sized`
+scales well to opting out of default bounds that have constness or hierarchies.
+
+This RFC's proposal to migrate from `?Sized` is based on ideas from an [earlier
+pre-eRFC][pre_erfc_fix_dsts] and then a [blog post][blog_dynsized_unsized] which developed
+on those ideas, and the feedback to this RFC's prior art, but is not a load-bearing part
+of this RFC.
+
+To preserve backwards compatibility, `Sized` bounds must be migrated to `const Sized` (see
+[the `size_of` and `size_of_val` section][size_of-and-size_of_val] for rationale), but
+`?Sized` bounds could retain their existing behaviour of removing the default `Sized` bound:
+
+### Keeping only `?Sized`
+[keeping-only-sized]: #keeping-only-sized
+
+Without adding any additional default bounds or relaxed forms, keeping `?Sized` could be
+compatible with this proposal as follows:
+
+In the current edition, `?Sized` would need to be equivalent to `?Sized + const ValueSized` to
+maintain backwards compatibility (see [the `size_of` and `size_of_val`
+section][size_of-and-size_of_val] for rationale). In the next edition, `?Sized` would be
+rewritten to `?Sized + const ValueSized` (remove the `Sized` default and add a `const ValueSized`
+bound) and bare `?Sized` would only remove the `Sized` default bound.
+
+Prior to the edition migration, the default bound is `Sized`, which could be changed using
+the following syntax:
+
+| With positive bounds | Keeping `?Sized` |
+| -------------------- | ---------------- |
+| `const Sized`        | `const Sized`    |
+| `Sized`              | `Sized`          |
+| `const ValueSized`   | `?Sized`         |
+| `ValueSized`         | Not possible     |
+| `Pointee`            | Not possible     |
+
+After the edition migration, the default bound is `const Sized`, which could be changed
+using the following syntax:
+
+| With positive bounds  | Keeping `?Sized`                  |
+| --------------------- | --------------------------------- |
+| `const Sized`         | `const Sized`                     |
+| `Sized`               | `?const Sized + Sized`            |
+| `const ValueSized`    | `?const Sized + const ValueSized` |
+| `ValueSized`          | `?const Sized + ValueSized`       |
+| `Pointee`             | `?const Sized`                    |
+
+### Adding `?ValueSized`
+[adding-valuesized]: #adding-valuesized
+
+Another alternative is to make `ValueSized` a default bound in addition to `Sized` and establish
+that relaxing a supertrait bound also implies relaxing subtrait bounds:
+
+Prior to the edition migration, the default bound is `ValueSized + const ValueSized + Sized`,
+which could be changed using:
```

Comment:
I am surprised by the lack of `const Sized` here. Currently code can call `size_of` in const context on all types, so `const Sized` should be part of the default bound, no?

Reply:
It should be, revisiting these tables, I felt they were a bit unclear so I've updated them to hopefully be a bit more understandable.

Reply:
Yes the new tables are much better, thanks :)